Source/Core/Core/AchievementManager.cpp

```diff
@@ -212,6 +279,70 @@ AchievementManager::ResponseType AchievementManager::FetchGameData()
       rc_api_process_fetch_game_data_response);
 }
 
+AchievementManager::ResponseType AchievementManager::FetchUnlockData(bool hardcore)
+{
+  rc_api_fetch_user_unlocks_response_t unlock_data{};
+  std::string username = Config::Get(Config::RA_USERNAME);
+  std::string api_token = Config::Get(Config::RA_API_TOKEN);
+  rc_api_fetch_user_unlocks_request_t fetch_unlocks_request = {.username = username.c_str(),
+                                                               .api_token = api_token.c_str(),
+                                                               .game_id = m_game_id,
+                                                               .hardcore = hardcore};
+  ResponseType r_type =
+      Request<rc_api_fetch_user_unlocks_request_t, rc_api_fetch_user_unlocks_response_t>(
+          fetch_unlocks_request, &unlock_data, rc_api_init_fetch_user_unlocks_request,
+          rc_api_process_fetch_user_unlocks_response);
+  if (r_type == ResponseType::SUCCESS)
+  {
+    m_lock.lock();
+    for (unsigned ix = 0; ix < unlock_data.num_achievement_ids; ix++)
+    {
+      m_unlock_map[unlock_data.achievement_ids[ix]].remote_unlock_status =
+          hardcore ? UnlockStatus::UnlockType::HARDCORE : UnlockStatus::UnlockType::SOFTCORE;
+      ActivateDeactivateAchievement(unlock_data.achievement_ids[ix]);
+    }
+    m_lock.unlock();
+  }
+  rc_api_destroy_fetch_user_unlocks_response(&unlock_data);
+  return r_type;
+}
+
+void AchievementManager::ActivateDeactivateAchievement(unsigned id)
+{
+  unsigned index = m_unlock_map[id].game_data_index;
+  bool active = (rc_runtime_get_achievement(&m_runtime, id) != nullptr);
+
+  // Deactivate achievements if game is not loaded
+  bool activate = m_is_game_loaded;
+  // Activate achievements only if achievements are enabled
+  if (activate && !Config::Get(Config::RA_ACHIEVEMENTS_ENABLED))
+    activate = false;
+  // Deactivate if achievement is unofficial, unless unofficial achievements are enabled
+  if (activate && !Config::Get(Config::RA_UNOFFICIAL_ENABLED) &&
+      m_game_data.achievements[index].category == RC_ACHIEVEMENT_CATEGORY_UNOFFICIAL)
+    activate = false;
+  // If encore mode is on, activate/deactivate regardless of current unlock status
+  if (activate && !Config::Get(Config::RA_ENCORE_ENABLED))
+  {
+    // Encore is off, achievement has been unlocked in this session, deactivate
+    activate = (m_unlock_map[id].session_unlock_count == 0);
```

Comment:
You should likely keep a reference to element retrieved from the map as a reference like:

```cpp
auto& unlock_status = m_unlock_map[id];
```

and then reuse it for all below occurrences in this block, otherwise you're going to be some unnecessary map lookups